### PR TITLE
Network module

### PR DESCRIPTION
**Pull request by sadym-chromium**
Start implementing [Network module](https://github.com/w3c/webdriver-bidi/pull/204).

### CODE REVIEW

File: tests/test_network.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import pytest
+from test_helpers import json, read_JSON_message, send_JSON_command, subscribe, \
+    execute_command
+
+
+@pytest.mark.asyncio
+async def test_network_before_request_sent_event_emitted(websocket, context_id):
+    await subscribe(websocket, "network.beforeRequestSent")
+
+    # document.cookie = 'foo=bar';
```
Comment (thiagowfx) [inline tests/test_network.py:25]: Add a TODO or remove

File: src/protocol/protocol.ts [hunk 2]
```diff
@@ -864,7 +938,8 @@ export namespace Session {
     | Log.EventNames
     | typeof Log.AllEvents
     | CDP.EventNames
-    | typeof CDP.AllEvents;
+    | typeof CDP.AllEvents
+    | Network.EventNames;
```
Comment (thiagowfx) [inline src/protocol/protocol.ts:942]: Missing:

```
| typeof Network.AllEvents
```

File: src/protocol/protocol.ts [hunk 3]
```diff
@@ -786,6 +790,76 @@ export namespace Log {
...
+    isRedirect: boolean;
+    navigation: BrowsingContext.Navigation | null;
+    redirectCount: number;
+    request: RequestData;
+    timestamp: number;
+  };
+
+  export type BeforeRequestSentParams = BaseEventParams & {};
+
+  export enum EventNames {
```
Comment (thiagowfx) [inline src/protocol/protocol.ts:858]: Missing: `AllEvents` definition.

Also ensure to update [`unrollEvents`](https://github.com/GoogleChromeLabs/chromium-bidi/blob/f777c628097e0a17664674a587f302c952c0a3cf/src/bidiMapper/domains/events/SubscriptionManager.ts#L42)

File: src/protocol-parser/protocol-parser.ts [hunk 4]
```diff
@@ -885,6 +955,7 @@ export namespace Session {
     Log.EventNames.LogEntryAddedEvent,
     CDP.AllEvents,
     CDP.EventNames.EventReceivedEvent,
+    Network.EventNames.BeforeRequestSentEvent,
```
Comment (thiagowfx) [inline src/protocol-parser/protocol-parser.ts:958]: Add `AllEvents` here

File: src/bidiMapper/domains/network/networkProcessor.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import Protocol from 'devtools-protocol';
+import {CdpClient} from '../../CdpConnection';
+import {IEventManager} from '../events/EventManager';
+import {LogType, LoggerFn} from '../../../utils/log';
+
+import {Network} from '../../../protocol/protocol.js';
+
+export class NetworkProcessor {
+  readonly #eventManager: IEventManager;
+  readonly #logger: LoggerFn | undefined;
```
Comment (thiagowfx) [inline src/bidiMapper/domains/network/networkProcessor.ts:27]: Prefer `readonly #logger?: LoggerFn;`

https://google.github.io/styleguide/tsguide.html#optionals-vs-undefined-type

File: src/bidiMapper/domains/network/networkProcessor.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+  static async init(
```
Comment (thiagowfx) [inline src/bidiMapper/domains/network/networkProcessor.ts:29]: For consistency with the rest of our codebase, prefer `create` instead of `init`

https://github.com/search?q=repo%3AGoogleChromeLabs%2Fchromium-bidi+static&type=code

**Comment by sadym-chromium** [comment]
It's a draft at very early stages

File: src/protocol/protocol.ts [hunk 7]
```diff
@@ -786,6 +790,117 @@ export namespace Log {
...
 
+export namespace Network {
+  export type Event = BeforeRequestSentEvent | ResponseCompletedEvent;
+
+  export type BeforeRequestSentEvent = EventResponse<
+    EventNames.BeforeRequestSentEvent,
+    BeforeRequestSentParams
+  >;
+
+  // TODO: implement.
```
Comment (thiagowfx) [inline src/protocol/protocol.ts:801]: Leftover TODO?

File: src/bidiMapper/domains/context/browsingContextImpl.ts [hunk 8]
```diff
@@ -140,8 +141,11 @@ export class BrowsingContextImpl {
     );
 
     // No need in awaiting for target to be unblocked.
+    // noinspection ES6MissingAwait
```
Comment (thiagowfx) [inline src/bidiMapper/domains/context/browsingContextImpl.ts:144]: This is from IntelliJ, right? It would be nice if you could either remove this annotation OR find something IDE-agnostic to express it. If this were python I am sure `# noqa` would work. For TypeScript there must be some other way, IDE-agnostic, to express "disable linting the next line".

File: runWPT.sh [hunk 9]
```diff
@@ -17,7 +17,6 @@ npm run build && \
   --webdriver-arg="--headless=true" \
   --log-wptreport wptreport.json \
   --manifest wpt/MANIFEST.json \
-  --metadata wpt-metadata/mapper/headless \
```
Comment (thiagowfx) [inline runWPT.sh:20]: Please revert the changes in this file before committing.

File: .eslintrc.js [hunk 10]
```diff
@@ -54,7 +54,6 @@ module.exports = {
     'no-underscore-dangle': 'error',
     'prefer-promise-reject-errors': 'error',
     'prefer-template': 'error',
-    'sort-imports': 'error',
```
Comment (thiagowfx) [inline .eslintrc.js:57]: Please revert the changes in this file before committing, especially in favor of https://github.com/GoogleChromeLabs/chromium-bidi/pull/523 which already does this.

File: src/bidiMapper/domains/network/networkProcessor.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+import Protocol from 'devtools-protocol';
+import {CdpClient} from '../../CdpConnection';
+import {IEventManager} from '../events/EventManager';
+
+import {NetworkRequest} from './networkRequest';
+
+export class NetworkProcessor {
+  readonly #eventManager: IEventManager;
+
+  readonly #requestMap = new Map<string, NetworkRequest>();
```
Comment (thiagowfx) [inline src/bidiMapper/domains/network/networkProcessor.ts:27]: Consider documenting this map (and use `/** */` when you do so).

File: src/bidiMapper/domains/network/networkRequest.ts [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+      ...baseEventParams,
+      initiator: {type: this.#getInitiatorType()},
+    };
+    return {
+      method: Network.EventNames.BeforeRequestSentEvent,
+      params,
+    };
+  }
+
+  #getInitiatorType(): 'parser' | 'script' | 'preflight' | 'other' {
```
Comment (thiagowfx) [inline src/bidiMapper/domains/network/networkRequest.ts:192]: Use `Network.Initiator.type` here (maybe you also need `typeof`)

File: src/bidiMapper/domains/network/networkRequest.ts [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+    switch (this.#requestWillBeSentEvent?.initiator.type) {
+      case 'parser':
+        return 'parser';
+      case 'script':
+        return 'script';
+      case 'preflight':
+        return 'preflight';
```
Comment (thiagowfx) [inline src/bidiMapper/domains/network/networkRequest.ts:199]: Optionally:

```
case 'parser':
case 'script':
case 'preflight':
  return this.#requestWillBeSentEvent?.initiator.type;
default:
  return 'other';
```

Only if you think it's more readable and/or less error-prone.

File: src/bidiMapper/domains/network/networkRequest.ts [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+      default:
+        return 'other';
+    }
+  }
+
+  static #getCookiesSameSite(
+    cdpSameSiteValue: string | undefined
+  ): 'strict' | 'lax' | 'none' {
```
Comment (thiagowfx) [inline src/bidiMapper/domains/network/networkRequest.ts:207]: Like the previous comment: Replace the type union with the proper definition in `protocol.ts`

File: src/bidiMapper/domains/context/browsingContextImpl.ts (hunk 8 above)
Comment (thiagowfx) [reply src/bidiMapper/domains/context/browsingContextImpl.ts:144]: I was doing some research, you can just prepend `void` to the expression, that should signal to the compiler there's no need to await for the promise.

Comment (thiagowfx) [reply src/bidiMapper/domains/context/browsingContextImpl.ts:144]: > Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator.eslint[@typescript-eslint/no-floating-promises](https://typescript-eslint.io/rules/no-floating-promises)

File: runWPT.sh (hunk 9 above)
Comment (sadym-chromium) [reply runWPT.sh:20]: Actually this script is not used in any of the CI actions, and is used for investigation. So I'd remove the metadata in order to have all the failure details.

Comment (thiagowfx) [reply runWPT.sh:20]: Given https://github.com/GoogleChromeLabs/chromium-bidi/pull/524, you can just use one of

```
HEADLESS=true ./runWPT.sh
HEADLESS=false ./runWPT.sh
```

...from now on.

File: src/bidiMapper/domains/network/networkProcessor.ts (hunk 11 above)
Comment (sadym-chromium) [reply src/bidiMapper/domains/network/networkProcessor.ts:28]: done

File: src/protocol/protocol.ts (hunk 7 above)
Comment (sadym-chromium) [reply src/protocol/protocol.ts:877]: done

File: runWPT.sh (hunk 9 above)
Comment (sadym-chromium) [reply runWPT.sh]: thanks!

**Comment by sadym-chromium** [comment]
Re-run CI

File: src/protocol-parser/protocol-parser.ts [hunk 15]
```diff
@@ -978,6 +1078,9 @@ export namespace Session {
     Log.EventNames.LogEntryAddedEvent,
     CDP.AllEvents,
     CDP.EventNames.EventReceivedEvent,
+    Network.AllEvents,
+    Network.EventNames.BeforeRequestSentEvent,
+    Network.EventNames.ResponseCompletedEvent,
```
Comment (thiagowfx) [inline src/protocol-parser/protocol-parser.ts:1083]: In the other file (`protocol.ts`), we have just `Network.EventNames` instead of listing each individual event. Can we do the same thing here? (same for the other event types, while we're here - http://go/tott/524)

File: src/bidiMapper/domains/events/EventManager.ts [hunk 16]
```diff
@@ -82,6 +84,7 @@ const eventBufferLength: ReadonlyMap<string, number> = new Map([
 ]);
 
 export class EventManager implements IEventManager {
+  readonly NETWORK_DOMAIN_PREFIX = 'network.';
```
Comment (thiagowfx) [inline src/bidiMapper/domains/events/EventManager.ts:87]: Is this meant to be an immutable constant for every instance?

If yes, please add `static` here as well

https://google.github.io/styleguide/tsguide.html#identifiers-constants

(Another possibility is to move it outside the class; your call.)

File: src/protocol-parser/protocol-parser.ts (hunk 15 above)
Comment (sadym-chromium) [reply src/protocol-parser/protocol-parser.ts:1083]: > In the other file (`protocol.ts`), we have just `Network.EventNames` instead of listing each individual event. Can we do the same thing here? (same for the other event types, while we're here - http://go/tott/524)

Out of scope of the change.

File: src/protocol-parser/protocol-parser.ts [hunk 17]
```diff
@@ -883,6 +883,106 @@ export namespace Log {
...
+
+  type BaseEventParams = {
+    context: CommonDataTypes.BrowsingContext | null;
+    navigation: BrowsingContext.Navigation | null;
+    redirectCount: number;
+    request: RequestData;
+    timestamp: number;
+  };
+
+  export type BeforeRequestSentParams = BaseEventParams & {};
```
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply src/protocol-parser/protocol-parser.ts]: Will be addressed by https://github.com/GoogleChromeLabs/chromium-bidi/pull/546

File: src/protocol-parser/protocol-parser.ts [hunk 18]
```diff
@@ -883,6 +883,106 @@ export namespace Log {
...
+    headers: Header[];
+    mimeType: string;
+    bytesReceived: number;
+    headersSize: number | null;
+    bodySize: number | null;
+    content: ResponseContent;
+  };
+
+  type BaseEventParams = {
+    context: CommonDataTypes.BrowsingContext | null;
```
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply src/protocol-parser/protocol-parser.ts]: I'd keep it close to the specification. Otherwise it would be hard to keep it up to date.

File: src/bidiMapper/domains/events/EventManager.ts (hunk 16 above)
Comment (sadym-chromium) [reply src/bidiMapper/domains/events/EventManager.ts]: done

File: src/protocol-parser/protocol-parser.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply src/protocol-parser/protocol-parser.ts]: UPD: extended the class

**Comment by sadym-chromium** [comment]
@thiagowfx  I reverted your merge commit, as it introduced unrelated changes in script evaluator.

File: src/protocol/protocol.ts (hunk 3 above)
Comment (sadym-chromium) [reply src/protocol/protocol.ts:973]: `AllEvents` is in line 971

Comment (sadym-chromium) [reply src/protocol/protocol.ts:973]: done
